File: ios/App/Sources/Core/Domain/NavCaddy/Memory/MissPatternStore.swift
import Foundation

/// Main interface for miss pattern storage and retrieval.
///
/// Spec R5: Miss Pattern Store tracks shot history and aggregates patterns
/// with time-based decay, filtering by club and pressure context.
///
/// This actor provides thread-safe access to miss pattern operations.
actor MissPatternStore {
    private let recorder: ShotRecorder
    private let aggregator: MissPatternAggregator
    private let repository: NavCaddyRepository

    init(
        recorder: ShotRecorder,
        aggregator: MissPatternAggregator,
        repository: NavCaddyRepository
    ) {
        self.recorder = recorder
        self.aggregator = aggregator
        self.repository = repository
    }

    /// Convenience initializer with repository dependency.
    init(repository: NavCaddyRepository) {
        let recorder = ShotRecorder(repository: repository)
        let aggregator = MissPatternAggregator(repository: repository)

        self.init(
            recorder: recorder,
            aggregator: aggregator,
            repository: repository
        )
    }

    // MARK: - Recording Misses

    /// Records a miss with full shot context.
    ///
    /// Spec Q3: Manual entry only for MVP.
    ///
    /// - Parameters:
    ///   - clubId: ID of the club used
    ///   - clubName: Name of the club (e.g., "7-iron")
    ///   - clubType: Type of club
    ///   - direction: Miss direction (push, pull, slice, hook, fat, thin)
    ///   - lie: Lie condition (tee, fairway, rough, etc.)
    ///   - pressure: Pressure context (user tagged or inferred)
    ///   - holeNumber: Optional hole number
    ///   - notes: Optional notes about the shot
    /// - Throws: Database error if recording fails
    func recordMiss(
        clubId: String,
        clubName: String,
        clubType: ClubType,
        direction: MissDirection,
        lie: Lie,
        pressure: PressureContext = PressureContext(),
        holeNumber: Int? = nil,
        notes: String? = nil
    ) async throws {
        let club = Club(
            id: clubId,
            name: clubName,
            type: clubType
        )

        let shot = Shot(
            timestamp: Date(),
            club: club,
            missDirection: direction,
            lie: lie,
            pressureContext: pressure,
            holeNumber: holeNumber,
            notes: notes
        )

        try await recorder.recordShot(shot)
    }

    /// Records a miss using an existing Club instance.
    ///
    /// - Parameters:
    ///   - club: The club used for the shot
    ///   - direction: Miss direction
    ///   - lie: Lie condition
    ///   - pressure: Pressure context
    ///   - holeNumber: Optional hole number
    ///   - notes: Optional notes
    /// - Throws: Database error if recording fails
    func recordMiss(
        club: Club,
        direction: MissDirection,
        lie: Lie,
        pressure: PressureContext = PressureContext(),
        holeNumber: Int? = nil,
        notes: String? = nil
    ) async throws {
        let shot = Shot(
            timestamp: Date(),
            club: club,
            missDirection: direction,
            lie: lie,
            pressureContext: pressure,
            holeNumber: holeNumber,
            notes: notes
        )

        try await recorder.recordShot(shot)
    }

    /// Records a successful shot (straight, no miss).
    ///
    /// - Parameters:
    ///   - club: The club used
    ///   - lie: Lie condition
    ///   - pressure: Pressure context
    ///   - holeNumber: Optional hole number
    ///   - notes: Optional notes
    /// - Throws: Database error if recording fails
    func recordSuccess(
        club: Club,
        lie: Lie,
        pressure: PressureContext = PressureContext(),
        holeNumber: Int? = nil,
        notes: String? = nil
    ) async throws {
        let shot = Shot(
            timestamp: Date(),
            club: club,
            missDirection: .straight,
            lie: lie,
            pressureContext: pressure,
            holeNumber: holeNumber,
            notes: notes
        )

        try await recorder.recordShot(shot)
    }

    // MARK: - Pattern Retrieval

    /// Gets miss patterns with optional filtering.
    ///
    /// - Parameters:
    ///   - clubId: Optional club ID to filter by
    ///   - pressure: Optional pressure context to filter by
    /// - Returns: Array of miss patterns sorted by confidence
    /// - Throws: Database error if fetch fails
    func getPatterns(
        clubId: String? = nil,
        pressure: PressureContext? = nil
    ) async throws -> [MissPattern] {
        if let clubId = clubId {
            return try await aggregator.getPatternsForClub(clubId)
        } else if let pressure = pressure, pressure.hasPressure {
            return try await aggregator.getPatternsForContext(pressure)
        } else {
            return try await aggregator.getAllPatterns()
        }
    }

    /// Gets patterns for a specific miss direction.
    ///
    /// - Parameter direction: The miss direction to filter by
    /// - Returns: Array of matching patterns
    /// - Throws: Database error if fetch fails
    func getPatternsByDirection(_ direction: MissDirection) async throws -> [MissPattern] {
        return try await aggregator.getPatternsByDirection(direction)
    }

    /// Gets the most significant patterns (highest confidence).
    ///
    /// - Parameter limit: Maximum number of patterns to return (default: 5)
    /// - Returns: Top patterns by confidence
    /// - Throws: Database error if fetch fails
    func getTopPatterns(limit: Int = 5) async throws -> [MissPattern] {
        let allPatterns = try await aggregator.getAllPatterns()
        return Array(allPatterns.prefix(limit))
    }

    /// Gets statistical summary of miss patterns.
    ///
    /// - Parameter windowDays: Rolling window in days (default: 30)
    /// - Returns: Pattern statistics
    /// - Throws: Database error if fetch fails
    func getStatistics(windowDays: Int = 30) async throws -> MissPatternAggregator.PatternStatistics {
        let shots = try await recorder.getShotsInWindow(days: windowDays)
        return aggregator.calculateStatistics(from: shots)
    }

    // MARK: - Shot History

    /// Gets recent shot history.
    ///
    /// - Parameter count: Maximum number of shots to return (default: 50)
    /// - Returns: Array of recent shots
    /// - Throws: Database error if fetch fails
    func getRecentShots(count: Int = 50) async throws -> [Shot] {
        return try await recorder.getRecentShots(count: count)
    }

    /// Gets shots for a specific club.
    ///
    /// - Parameters:
    ///   - clubId: The club ID
    ///   - limit: Maximum number of shots (default: 50)
    /// - Returns: Array of shots for the club
    /// - Throws: Database error if fetch fails
    func getShotsByClub(_ clubId: String, limit: Int = 50) async throws -> [Shot] {
        return try await recorder.getShotsByClub(clubId, limit: limit)
    }

    /// Gets shots that occurred under pressure.
    ///
    /// - Parameter limit: Maximum number of shots (default: 50)
    /// - Returns: Array of pressure shots
    /// - Throws: Database error if fetch fails
    func getPressureShots(limit: Int = 50) async throws -> [Shot] {
        return try await recorder.getShotsWithPressure(limit: limit)
    }

    // MARK: - Memory Management

    /// Clears all shot history and patterns.
    ///
    /// Spec C4: User-facing control to wipe memory.
    ///
    /// - Throws: Database error if clear fails
    func clearHistory() async throws {
        try await repository.clearMemory()
    }

    /// Enforces the retention policy by deleting old data.
    ///
    /// Spec Q5: 90-day retention window.
    ///
    /// - Throws: Database error if cleanup fails
    func enforceRetentionPolicy() async throws {
        try await repository.enforceRetentionPolicy()
    }
}

// MARK: - Convenience Extensions

extension MissPatternStore {
    /// Checks if a pattern exists for a specific club and direction.
    ///
    /// - Parameters:
    ///   - clubId: The club ID
    ///   - direction: The miss direction
    /// - Returns: True if pattern exists with confidence > 0.3
    /// - Throws: Database error if fetch fails
    func hasSignificantPattern(clubId: String, direction: MissDirection) async throws -> Bool {
        let patterns = try await getPatterns(clubId: clubId)
        return patterns.contains { $0.direction == direction && $0.confidence > 0.3 }
    }

    /// Gets a human-readable summary of miss patterns for a club.
    ///
    /// - Parameter clubId: The club ID
    ///   - Returns: Textual summary of patterns
    /// - Throws: Database error if fetch fails
    func getPatternSummary(clubId: String) async throws -> String {
        let patterns = try await getPatterns(clubId: clubId)

        guard !patterns.isEmpty else {
            return "No miss patterns found for this club."
        }

        let topPattern = patterns[0]
        let confidence = Int(topPattern.confidence * 100)

        return "Most common miss: \(topPattern.direction.rawValue) (\(confidence)% confidence)"
    }
}

File: android/app/src/main/java/com/example/app/domain/navcaddy/memory/MissPatternStore.kt
package caddypro.domain.navcaddy.memory

import caddypro.domain.navcaddy.models.Club
import caddypro.domain.navcaddy.models.Lie
import caddypro.domain.navcaddy.models.MissDirection
import caddypro.domain.navcaddy.models.MissPattern
import caddypro.domain.navcaddy.models.PressureContext
import kotlinx.coroutines.flow.Flow
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Main facade for the Miss Pattern Store subsystem.
 *
 * Combines shot recording and pattern aggregation into a unified interface
 * for the NavCaddy memory system.
 *
 * Spec reference: navcaddy-engine.md R5 (Miss Pattern Memory)
 */
@Singleton
class MissPatternStore @Inject constructor(
    private val shotRecorder: ShotRecorder,
    private val patternAggregator: MissPatternAggregator
) {
    // ========================================================================
    // Shot Recording
    // ========================================================================

    /**
     * Record a miss with full context.
     *
     * Primary method for manual miss entry (Q3: manual only for MVP).
     *
     * @param club Club used for the shot
     * @param direction Direction of miss
     * @param lie Ball lie before the shot
     * @param pressure Pressure context for the shot
     * @param holeNumber Hole number if shot was during a round (optional)
     * @param notes Optional user notes about the shot
     */
    suspend fun recordMiss(
        club: Club,
        direction: MissDirection,
        lie: Lie,
        pressure: PressureContext,
        holeNumber: Int? = null,
        notes: String? = null
    ) {
        shotRecorder.recordShot(
            club = club,
            missDirection = direction,
            lie = lie,
            pressureContext = pressure,
            holeNumber = holeNumber,
            notes = notes
        )
    }

    /**
     * Record a miss with simplified parameters.
     *
     * Convenience method for basic miss entry.
     *
     * @param clubId ID of the club used
     * @param clubName Display name of the club
     * @param direction Direction of miss
     * @param lie Ball lie before the shot
     * @param isUserTaggedPressure Whether user marked this as a pressure shot
     */
    suspend fun recordMiss(
        clubId: String,
        clubName: String,
        direction: MissDirection,
        lie: Lie,
        isUserTaggedPressure: Boolean = false
    ) {
        val clubType = inferClubType(clubName)
        val club = Club(
            id = clubId,
            name = clubName,
            type = clubType,
            estimatedCarry = inferDefaultCarry(clubType)
        )

        val pressure = PressureContext(
            isUserTagged = isUserTaggedPressure,
            isInferred = false
        )

        recordMiss(
            club = club,
            direction = direction,
            lie = lie,
            pressure = pressure
        )
    }

    // ========================================================================
    // Pattern Retrieval
    // ========================================================================

    /**
     * Get all current patterns with decay applied.
     *
     * Returns patterns from recent shot history (default 30 days, 50 shots).
     *
     * @param clubId Optional club ID to filter patterns (null = all clubs)
     * @param pressure Optional pressure context to filter patterns (null = all contexts)
     * @return List of miss patterns ordered by confidence descending
     */
    suspend fun getPatterns(
        clubId: String? = null,
        pressure: PressureContext? = null
    ): List<MissPattern> {
        return when {
            // Club + pressure filter
            clubId != null && pressure?.hasPressure == true -> {
                patternAggregator.getPatternsForClubAndPressure(clubId)
            }

            // Club filter only
            clubId != null -> {
                patternAggregator.getPatternsForClub(clubId)
            }

            // Pressure filter only
            pressure?.hasPressure == true -> {
                patternAggregator.getPatternsForPressure()
            }

            // No filters - all patterns
            else -> {
                patternAggregator.aggregatePatterns()
            }
        }
    }

    /**
     * Get patterns as a Flow for reactive updates.
     *
     * Subscribes to pattern changes with real-time decay applied.
     *
     * @return Flow of patterns ordered by decayed confidence
     */
    fun getPatternsFlow(): Flow<List<MissPattern>> {
        return patternAggregator.getPatternsWithDecay()
    }

    /**
     * Get patterns for a specific club.
     *
     * @param clubId The club ID to filter by
     * @return List of club-specific patterns
     */
    suspend fun getPatternsForClub(clubId: String): List<MissPattern> {
        return getPatterns(clubId = clubId)
    }

    /**
     * Get patterns for pressure situations.
     *
     * @return List of pressure-specific patterns
     */
    suspend fun getPatternsForPressure(): List<MissPattern> {
        return getPatterns(
            pressure = PressureContext(isUserTagged = true, isInferred = true)
        )
    }

    /**
     * Get the most dominant pattern (highest confidence).
     *
     * @param clubId Optional club ID to filter by
     * @param pressure Optional pressure context to filter by
     * @return Most dominant pattern or null if no patterns exist
     */
    suspend fun getDominantPattern(
        clubId: String? = null,
        pressure: PressureContext? = null
    ): MissPattern? {
        return getPatterns(clubId, pressure).maxByOrNull { it.confidence }
    }

    // ========================================================================
    // Shot History
    // ========================================================================

    /**
     * Get recent shot history.
     *
     * @param count Maximum number of shots to retrieve
     * @return List of recent shots ordered by timestamp descending
     */
    suspend fun getRecentShots(count: Int): List<caddypro.domain.navcaddy.models.Shot> {
        return shotRecorder.getMostRecentShots(count)
    }

    /**
     * Get shot history for a specific club.
     *
     * @param clubId The club ID to filter by
     * @param limit Maximum number of shots to retrieve
     * @return List of club shots ordered by timestamp descending
     */
    suspend fun getShotsByClub(clubId: String, limit: Int): List<caddypro.domain.navcaddy.models.Shot> {
        return shotRecorder.getShotsByClub(clubId, limit)
    }

    // ========================================================================
    // Memory Management
    // ========================================================================

    /**
     * Clear all miss history and patterns.
     *
     * User-requested memory wipe per C4.
     *
     * Spec reference: navcaddy-engine.md C4 (privacy controls)
     */
    suspend fun clearHistory() {
        shotRecorder.clearHistory()
    }

    /**
     * Enforce retention policy and clean up old data.
     *
     * Deletes shots and patterns outside the 90-day retention window.
     *
     * @return Number of items deleted
     */
    suspend fun enforceRetentionPolicy(): Int {
        val shotsDeleted = shotRecorder.enforceRetentionPolicy()
        val patternsDeleted = patternAggregator.clearStalePatterns()
        return shotsDeleted + patternsDeleted
    }

    /**
     * Refresh patterns from current shot history.
     *
     * Re-aggregates patterns and persists to database.
     * Useful for periodic pattern updates or manual refresh.
     *
     * @return Number of patterns updated
     */
    suspend fun refreshPatterns(): Int {
        val patterns = patternAggregator.aggregatePatterns()
        patternAggregator.savePatterns(patterns)
        return patterns.size
    }

    // ========================================================================
    // Helper Methods
    // ========================================================================

    /**
     * Infer club type from club name.
     *
     * Simple heuristic for club type inference when creating clubs from strings.
     * Not comprehensive - for production, use a proper club database.
     *
     * @param clubName Display name of the club
     * @return Inferred club type
     */
    private fun inferClubType(clubName: String): caddypro.domain.navcaddy.models.ClubType {
        val name = clubName.lowercase()
        return when {
            "driver" in name || "1w" in name -> caddypro.domain.navcaddy.models.ClubType.DRIVER
            "wood" in name || name.matches(Regex("\\d+w")) -> caddypro.domain.navcaddy.models.ClubType.WOOD
            "hybrid" in name || name.matches(Regex("\\d+h")) -> caddypro.domain.navcaddy.models.ClubType.HYBRID
            "iron" in name || name.matches(Regex("[3-9]i?")) -> caddypro.domain.navcaddy.models.ClubType.IRON
            "wedge" in name || "pw" in name || "sw" in name || "lw" in name || "gw" in name -> caddypro.domain.navcaddy.models.ClubType.WEDGE
            "putter" in name -> caddypro.domain.navcaddy.models.ClubType.PUTTER
            else -> caddypro.domain.navcaddy.models.ClubType.IRON // Default fallback
        }
    }

    /**
     * Infer default carry distance from club type.
     *
     * Provides reasonable default carry distances in meters for each club type.
     * Used when creating clubs from strings without explicit distance info.
     *
     * @param clubType The club type
     * @return Estimated carry distance in meters
     */
    private fun inferDefaultCarry(clubType: caddypro.domain.navcaddy.models.ClubType): Int {
        return when (clubType) {
            caddypro.domain.navcaddy.models.ClubType.DRIVER -> 220
            caddypro.domain.navcaddy.models.ClubType.WOOD -> 190
            caddypro.domain.navcaddy.models.ClubType.HYBRID -> 170
            caddypro.domain.navcaddy.models.ClubType.IRON -> 140
            caddypro.domain.navcaddy.models.ClubType.WEDGE -> 100
            caddypro.domain.navcaddy.models.ClubType.PUTTER -> 0
        }
    }
}
